man/etherwake.1
\." $Id$
.\"
.TH ETHERWAKE 1 "Release 0.1.3" "LLNL" "ETHERWAKE"

.SH NAME
etherwake \- low-level power control

.SH SYNOPSIS
.B etherwake
.I "[-options] [command]"

.SH DESCRIPTION
.B etherwake
is a low level utility in support of the PowerMan power management 
program.  It communicates with a target node via the ether-wake
"magic packet".  
For a command in 
.I [on, off, reset]
.B etherwake
will send a magic packet to the target nodes.
.B etherwake
conforms to the PowerMan power monitoring interface.  One will generally
prefer to use 
.B pm
rather than this low-level utility, but its use is not forbidden to 
root priveleged users.  Be aware
that ethernet wake-on-LAN is implemented (in some cases) as a TOGGLE.  If 
you send a wake-up to such a node it will shut down.  This could be bad.
Options are described in TARGET NODE LIST OPTIONS and OTHER OPTIONS below.
.LP
The POWERMAN environment variable, if set, identifies an alternate 
directory fo the PowerMan libraries and configuration files.

.SH TARGET NODE LIST OPTIONS
.I "-w host,host,..."
Target the specified nodes.  No spaces are allowed in the comma-separated
list.  Substituting the "-" character for the list causes the target nodes
to be read from stdin, one per line.
.LP
.I "-a"
Target "all" nodes in the cluster.  
.B etherwake
determines the list of nodes in the cluster from the configuration file
(see FILES below).

.SH OTHER OPTIONS
.LP
.I "-c configfile"
Provide an alternate name or location for the configuration file.
.LP
.I "-f number"
Set the maximum concurrency, currently unused.  
.I "-l powermandir"
Provide an alternate location for the PowerMan libraries.
.LP
.I "-v"
be verbose, i.e. print error messages rather than dying quietly.

.SH "FILES"
$powermandir/bin/etherwake
.br
$powermandir/etc/etherwake.conf
.br
$powermandir/lib/ether-wake
.br

.SH "ORIGIN"
Developed by Andrew  Uselton <useton2@llnl.gov> on LLNL's Alpha and
Intel Linux clusters.

.SH "SEE ALSO"
digi(1), etherwake.conf(5), pm(1), rmc(1), wti(1)



NEWS
=========================================================================
Release Notes for PowerMan version 2.4                        26 Nov 2008
=========================================================================

* Initial powerman client API.

* Run the powerman daemon (and all coprocesses) as 'daemon' by default
  instead of root.  To override, set USER=root or other user in
  /etc/sysconfig/powerman.

* New man pages for httppower, plmpower.

=========================================================================
Release Notes for PowerMan version 2.3                        12 Nov 2008
=========================================================================

* Make init script work on Solaris.

* IPv6 support.

* Include example powerman.conf file (/etc/powerman/powerman.conf.example)

* Added support for Cyclades PM20, PM42.

=========================================================================
Release Notes for PowerMan version 2.2                        27 Aug 2008
=========================================================================

* Send a SIGINT to coprocesses rather than just closing file
  descriptors during powermand shutdown.  This is required to 
  terminate an ssh client running in coprocess mode.

* Make it possible to only define the _ranged version of scripts.  Scripts 
  are selected using the following precedence: _all, _ranged, singlet.

* Drop "soft power status" support.  This allowed iceboxes to detect 
  whether a node was powered up or not, independently of plug state.

* [ipmipower] Drop _all and singlet version of scripts. 

* [ilom] Revised Sun Integrated Lights Out Management support 
  to work over ssh and serial.  Dropped shared console/serial 
  "ilom-inline" support.

* [icebox3] Now supports both v3 and v4 Ice Box.  Users of icebox4
  should change their device types to icebox3 in powerman.conf.

* [plmpower] Added support for controlling Insteon/X10 devices via 
  SmartLabs PLM 2412S.

* [hp3488] Added support for modular GPIB-based HP 3488A switch/control 
  unit via gpib-utils project.

* [ics8064] Added support for VXI-11-based ICS 8064 16-port relay unit
  via gpib-utils project.

* [powerman] Improved efficiency of powerman-controlling-powerman 
  configurations, when one powerman controls a subset of the plugs of 
  another powerman through the use of _ranged scripts.

* Enhanced integrated test suite.
  Note: tests all pass on AIX now.

=========================================================================
Release Notes for PowerMan version 2.1                        18 Jun 2008
=========================================================================

* Client overhaul.

* [apcpdu3] Fix intermittent query failures.

* Fixed off by one bug in server that allowed some delays in dev 
  scripts to take longer than programmed.

* Use raw mode in pipe (|&) connections to avoid local tty echo.

* Run server in /var/run/powerman rather than /etc/powerman.
  Create this directory if it doesn't exist.

* Enhanced integrated test suite.
  Note: all tests do not currently pass on AIX.

=========================================================================
Release Notes for PowerMan version 2.0                        01 Jun 2008
=========================================================================

* Autoconf/automake integration.

* Integrated test suite.

* Portability to Solaris, AIX, and OS/X.

* Support for selecting power control targets using genders (-g option).

* Cleanup and refactoring.

=========================================================================
Release Notes for PowerMan version 1.0.32                     24 Jan 2008
=========================================================================

* Support for new apc firmware (apcpdu3.dev) [Trent D'Hooge]

* Added httppower utility for interfacing to power controllers that are
  exclusively web-based.

* Support for Digital Loggers, Inc. power controllers.

* Support for APPRO power controller.

=========================================================================
Release Notes for PowerMan version 1.0.31                     20 Oct 2007
=========================================================================

(Includes changes in 1.0.29 and 1.0.30)

* Support SUN ILOM inline with serial console, i.e.
  device "ilom25" "ilom-inline" "/usr/bin/conman -Q -j ilc25 |&"

* Handle config files with embedded carriage returns [Trent D'Hooge
  and Todd Heer].

* Ipmipower device script tuning [Al Chu].

* Minor build/packaging tweaks for building under mock, etc.

=========================================================================
Release Notes for PowerMan version 1.0.28                     07 May 2007
=========================================================================

* Increase "cycle" delays from 2s to 4s on all devices that implement
  cycle as an explicit on,delay,off.

=========================================================================
Release Notes for PowerMan version 1.0.27                     24 Apr 2007
=========================================================================

* Add heartbeat stonith module.

* Fix bug in baytech-rpc28-nc device support that affected plugs > 10.

=========================================================================
Release Notes for PowerMan version 1.0.26                     21 Dec 2006
=========================================================================

* Support 8-port APC AP7900 and likely AP7901, AP7920, and AP7921.
  [Martin Petersen]

* Work around a bug that causes baytech rpc3's to resturn some plugs
  as status "unknown" on login timeout/reconnect.

* Add ranged power control support for faster power control w/ ipmipower.

=========================================================================
Release Notes for PowerMan version 1.0.25                     16 Aug 2006
=========================================================================


lib/etherwake.tcl
####################################################################
# $Id$
# by Andrew C. Uselton <uselton2@llnl.gov> 
# Copyright (C) 2000 Regents of the University of California
# See ../DISCLAIMER
# v. 0-0-1:  2001-07-11
# v. 0-0-2:  2001-07-31
####################################################################

####################################################################
# power control type:
# etherwake
####################################################################

namespace eval etherwake {

proc power {command node_list} {
    set dir ""
    if { $command == "on" } {
        set dir "off"
    }
    if { $command == "off" } {
        set dir "on"
    }
    if { $command == "reset" } {
        set dir "both"
	::powerlib::etherwake::power "off" $node_list
	if {$::app::tkdanger} {
	    set ::powerlib::etherwake::delay_list $node_list
	    after 3000 {
		::powerlib::etherwake::power "on" $::etherwake::delay_list
	    }
	} else {
	    after 3000 
	    ::powerlib::etherwake::power "on" $node_list
	}
	return
    }
    if { $dir == "" } {
        puts "etherwake::power does not understand $command"
    } else {
	foreach node $node_list {
	    if {[llength [::powerlib::check $node $dir]] > 0} {	
		set mac_addr [lindex $::powerlib::control($node) 1]
		set etherwake_command [format "exec %s/ether-wake %s" $::app::lib_dir $mac_addr]
		eval $etherwake_command
	    }
	}
    }
}

}